### PR TITLE
feat(#major); ENS; Add ENS

### PR DESCRIPTION
**Pull request by harsh9200**
Added ENS substreams to index all ENS mappings and reverse mappings.

### CODE REVIEW

**Comment by this-username-is-taken** [comment]
Now that Robin fixed the build, you should look into why it's failing. We should try to ensure every PR builds going forward.

File: ens/README.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# ENS
+
+NOTE: At the time of writing, the ens substream only works for ENS registered with ETH Base Registrar contract.
```
Comment (this-username-is-taken) [inline ens/README.md:3]: You should add some description on what this module is for. Would also be helpful to add it into the PR description section

File: Cargo.toml [hunk 2]
```diff
@@ -11,6 +11,7 @@ members = [
     "substreams-helper",
     "uniswap-v2",
     "solana-sample",
+    "ens",
```
Comment (this-username-is-taken) [inline Cargo.toml:14]: I'd rename this as `ens-lookup` as ENS can mean a lot of things (e.g. could be a substream for ENS the protocol)

File: ens-names/substreams.yaml [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    - ens.proto
+    - common.proto
+  importPaths:
+    - ./proto/v1
+    - ../common/proto
+
+binaries:
+  default:
+    type: wasm/rust-v1
+    file: "../target/wasm32-unknown-unknown/release/ens.wasm"
```
Comment (this-username-is-taken) [inline ens-names/substreams.yaml:20]: should be `ens_names.wasm`

File: ens-names/proto/v1/ens.proto [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    string ens_name = 1;
+    string name_hash = 2;
+    
+    string label_name = 3;
+    string label_hash = 4;
+
+    string controller_address = 5;
+    string registrant_address = 6;
+
+    string transaction_hash = 7;
```
Comment (this-username-is-taken) [inline ens-names/proto/v1/ens.proto:15]: can we also add `block_number` here?

File: ens-names/substreams.yaml (hunk 3 above)
Comment (harsh9200) [reply ens-names/substreams.yaml:20]: nice catch!

File: ens-names/proto/v1/ens.proto (hunk 4 above)
Comment (harsh9200) [reply ens-names/proto/v1/ens.proto:15]: Sure

**Comment by this-username-is-taken** [comment]
what happens when the controller changes? do we update the store?

File: ens-names/src/keyer.rs [hunk 5]
```diff
@@ -0,0 +1,7 @@
+pub fn ens_domain_key(asset_address: &String) -> String {
+    format!("domain:{}", asset_address)
```
Comment (this-username-is-taken) [inline ens-names/src/keyer.rs:2]: what does an `asset_address` refer to here?

File: ens-names/README.md [hunk 6]
```diff
@@ -0,0 +1,3 @@
+# ENS Names
+
+NOTE: At the time of writing, the ens substream only works for ENS registered with ETH Base Registrar contract.
```
Comment (this-username-is-taken) [inline ens-names/README.md:3]: you should add a more detailed description about what this substream does

Comment (harsh9200) [reply ens-names/README.md]: Sure

File: ens-names/src/keyer.rs (hunk 5 above)
Comment (harsh9200) [reply ens-names/src/keyer.rs]: I have renamed this to `ens_hash`.

**Review by this-username-is-taken — APPROVED** [review]
Looks great!

**Pull request merged**